### DISCUSSION TITLE
Plot - y-axis autoscale and scientific notation

### DISCUSSION THREAD

**Discussion by s-celles** [OpenModelica/OpenModelica#11200] [Q&A] [answered]
Hello,

sometimes y-axis scale doesn't autoscale very well

![image](https://github.com/OpenModelica/OpenModelica/assets/109167/1c27fc34-bf62-47b1-b533-78efd9db9ae8)

![image](https://github.com/OpenModelica/OpenModelica/assets/109167/329bc201-9e57-4289-b7bd-77390f9a4eec)

Currently it seems that there is a conversion with engineering notation (10^3, 10^6 ...)

I wonder if using scientific notation (of numbers) couldn't help to have better autoscale.

In such an example force is going from -62.5kN to 62.5 kN and speed from -3 m/s to 3 m/s

You should notice that km/s is really not an appropriate unit here !

Using scientific notation could help because torque could use axis from -6.25 to 6.25 (10^4 N.m) and speed -3 to 3 (m/s)

Any opinion ?

**Comment by s-celles** [comment]
Maybe a solution could be to have a settings in Plot Setup / Range

![image](https://github.com/OpenModelica/OpenModelica/assets/109167/49366ab3-bdd6-452f-82a1-803b8c71267a)

instead of a Checkbox "Prefix Units" a dropdown with

- Engineering notation - Prefix Units (default because this is the current behaviour)
- Scientific notation - Prefix Units   (y axis will go in such an example from -8 to 8)
- Scientific notation - No Prefix Units (y axis will go in such an example from -8e+04 to 8e+04)  (the current behaviour when actual Checkbox "Prefix Units" is unchecked

I would move this dropdown before "Auto Scale" checkbox because that could become a more important setting

When Scientific notation - Prefix Units is choosed I will display as legend

forceSensor_chariot.f (10^4 N)     speed.v (m/s)

**Comment by casella** [comment]
The speed in km/s at the tooltip is a bug. Please open a ticket about it.

Regarding the other issue, I can't see the point of having scientific notatio with prefixes. The whole point of using prefixes is to have numbers which are neither too large nor too small - why bother about scientific notation in that case?

BTW, although it is not forbidden to do so, plotting two variables that have different quantities in the same plot is not a good idea. On one hand, you can't compare them (can't compare apples and pears). On the other hand, it is very likely that the scale required for the largest variable will make the plot of the other one very close to the origin.

**Reply by s-celles** [reply]
Using scientific notation without prefix will help to have all values between -10 and 10 and so have them with visible differences 
That's not for comparing their value but to compare the shape / form of their plot.

**Reply by s-celles** [reply]
The main problem is not using scientific notation or not... because when disabling Checkbox "Prefix Units", OpenModelica uses scientific notation as this screeenshot can attest 
![image](https://github.com/OpenModelica/OpenModelica/assets/109167/5d99c34a-6ddf-452e-8847-a73f789a3bd9)
The problem is, when using scientific notation, hidding or not power of 10 ie having in such a case axis from -8 to 8 instead of -8e+04 to 8e+04.

**Reply by bilderbuchi** [reply]
> BTW, although it is not forbidden to do so, plotting two variables that have different quantities in the same plot is not a good idea. On one hand, you can't compare them (can't compare apples and pears).

That seems like a very weird assertion to me? In my experience half the analysis I do on ODE simulation results is how changes in one variable interact with another variable. 

>  On the other hand, it is very likely that the scale required for the largest variable will make the plot of the other one very close to the origin.

Plotting variables together is how I do the above analysis (of course?) -- the plotting system should help me by making it easy to avoid the problem of different scales! Liberal use of `displayUnit` can help with this, but good prefix scaling seems important, too.

**Comment by s-celles** [comment]
Here is an example of what such a plotting y-axis autoscale could provide

![image](https://github.com/OpenModelica/OpenModelica/assets/109167/36c8443e-e7b9-44cf-8e39-0bb9eaaa72ac)

@casella You can see that it's much easier now to explain what is going on the two variables.

Phase 1 : Force is positive / speed is increasing linearly
Phase 2 : Force is equal to zero / speed is a positive constant (3 m/s)
Phase 3 : Force is negative / speed is decreasing linearly up to 0
Phase 4 : Force is negative / speed is going more and more negative linearly
Phase 5 : Force is equal to zero / speed is a negative constant (-3 m/s)
Phase 6 : Force is positivie / speed is negative but increasing linearly to zero
Phase 7 : Force is equal to zero / speed is equal to 0

This is the kind of analysis @bilderbuchi is probably mentioning.

Sometimes (often) that can be very useful.

Here is model to reproduce (but not currently with the kind of plotting OpenModelica 1.21.0 provides) :

```modelica
model chariot_portique
  constant Real rpm_max = 1450.0;
  constant Real v_max = 3.0;
  
  Modelica.Mechanics.Rotational.Components.IdealGearR2T idealGearR2T(ratio = (rpm_max*2*Modelica.Constants.pi/60)/v_max)  annotation(
    Placement(visible = true, transformation(origin = {28, -36}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Sensors.SpeedSensor speedSensor_moteur annotation(
    Placement(visible = true, transformation(origin = {-52, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Components.LossyGear lossyGear(lossTable = [0, 0.8, 0.8, 0, 0], ratio = 1)  annotation(
    Placement(visible = true, transformation(origin = {-6, -36}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Translational.Sensors.SpeedSensor speedSensor_chariot annotation(
    Placement(visible = true, transformation(origin = {72, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Translational.Components.Mass mass(m = 82000)  annotation(
    Placement(visible = true, transformation(origin = {122, -36}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Sources.CombiTimeTable combiTimeTable(extrapolation = Modelica.Blocks.Types.Extrapolation.HoldLastPoint, table = [0, 0; 0, 1520; 4, 1520; 4, 0; 8, 0; 8, -974; 12, -974; 12, 0; 16, 0; 16, -1520; 20, -1520; 20, 0; 24, 0; 24, 974; 28, 974; 28, 0; 32, 0], tableOnFile = false) annotation(
    Placement(visible = true, transformation(origin = {-66, -66}, extent = {{-80, 20}, {-60, 40}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Sources.Torque torque annotation(
    Placement(visible = true, transformation(origin = {-100, -36}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Translational.Sensors.ForceSensor forceSensor_chariot annotation(
    Placement(visible = true, transformation(origin = {90, -36}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Translational.Sensors.PowerSensor powerSensor_chariot annotation(
    Placement(visible = true, transformation(origin = {60, -36}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Sensors.PowerSensor powerSensor_moteur annotation(
    Placement(visible = true, transformation(origin = {-68, -36}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Sensors.TorqueSensor torqueSensor_moteur annotation(
    Placement(visible = true, transformation(origin = {-36, -36}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(lossyGear.flange_b, idealGearR2T.flangeR) annotation(
    Line(points = {{4, -36}, {18, -36}}));
  connect(speedSensor_chariot.flange, idealGearR2T.flangeT) annotation(
    Line(points = {{62, 0}, {38, 0}, {38, -36}}, color = {0, 127, 0}));
  connect(combiTimeTable.y[1], torque.tau) annotation(
    Line(points = {{-125, -36}, {-112, -36}}, color = {0, 0, 127}));
  connect(forceSensor_chariot.flange_b, mass.flange_a) annotation(
    Line(points = {{100, -36}, {112, -36}}, color = {0, 127, 0}));
  connect(powerSensor_chariot.flange_a, idealGearR2T.flangeT) annotation(
    Line(points = {{50, -36}, {38, -36}}, color = {0, 127, 0}));
  connect(powerSensor_chariot.flange_b, forceSensor_chariot.flange_a) annotation(
    Line(points = {{70, -36}, {80, -36}}, color = {0, 127, 0}));
  connect(torque.flange, powerSensor_moteur.flange_a) annotation(
    Line(points = {{-90, -36}, {-78, -36}}));
  connect(powerSensor_moteur.flange_b, torqueSensor_moteur.flange_a) annotation(
    Line(points = {{-58, -36}, {-46, -36}}));
  connect(lossyGear.flange_a, torqueSensor_moteur.flange_b) annotation(
    Line(points = {{-16, -36}, {-26, -36}}));
  connect(speedSensor_moteur.flange, torque.flange) annotation(
    Line(points = {{-62, 0}, {-90, 0}, {-90, -36}}));
  annotation(
    experiment(StopTime = 32),
    uses(Modelica(version = "4.0.0")),
  Diagram(coordinateSystem(extent = {{-200, -100}, {200, 100}})),
  Icon(coordinateSystem(extent = {{-200, -100}, {200, 100}})),
  version = "");
end chariot_portique;

```

**Comment by casella** [comment]
> > BTW, although it is not forbidden to do so, plotting two variables that have different quantities in the same plot is not a good idea. On one hand, you can't compare them (can't compare apples and pears).
> 
> That seems like a very weird assertion to me? In my experience half the analysis I do on ODE simulation results is how changes in one variable interact with another variable.

Don't get me wrong, I'm a control engineer, so the main reason why I use Modelica is to figure out how input variables influence output variables; seeing them in the same plot is essential to see if there are delays, which are detrimental to feedback control performance, particularly when you zoom in. What I meant was: plotting them on the same plot _with the same Y grid_.

The problem is, trying to get the two plots to have more or less the same vertical extent by playing around with displayUnits is the poor man's solution for this. If you want to compare _two_ variables (or two sets of variables) with different numerical value extents, the proper way to do that is to have two Y grids, one at the left and one at the right. We could implement this feature in OMEdit and it would be very useful.

The other way (which I always do with control system analysis) is to normalize them appropriately , so that all variables have a range -1 to 1 and can nicely fit the same plot without the need of multiple Y grids. 

Having a button to do this automatically, using max(abs(v), v.nominal) as a scaling value, would be very useful to me. The actual numerical values could be enquired with the tooltip.

@bilderbuchi, @scls19fr what do you think?

**Reply by bilderbuchi** [reply]
> have two Y grids, one at the left and one at the right. We could implement this feature in OMEdit and it would be very useful.

This sounds a very useful feature in itself (with e.g. a selector in the plot menu which axis to attach a given trace to)! Having that will probably alleviate an appreciable fraction of the discussed problems.

Normalising seems attractive, but you lose a lot of expressiveness that way, and any exports are heavily reduced in usefulness -- you don't have a tooltip in a png, and keeping track of the respective scaling values e.g. in the legend will consume a lot of space. 

In my experience, using appropriate `displayUnit` and having engineering prefixes working correctly is "good enough" for most cases.
Sometimes I'm frustrated when I do this correctly and then need to plot a "foreign" (i.e. not in my code) variable which has the wrong displayUnit. For example, I have some displacements in mm magnitude, and I plot those with some pressures in bar magnitude -- works out nicely and displays well, but then I need to add some components `flange.s`, which is in metre and not selectable otherwise -- enabling a user to manually select "milli" would help with this.

**Reply by bilderbuchi** [reply]
Another thing: It should always be possible to meaningfully plot _any_ quantity of _arbitrary_ magnitude in engineering prefixes, as long as one uses two plots/axes: 
* values are scaled to the nearest engineering prefix, so generally vary from 1-1000. This is too much span to display in one axis, so
* traces are assigned to a "high" or a "low" bucket, where one contains value magnitudes 30-1000 and the other 1-30 (well `sqrt(1000)` but round numbers are nice, and maybe extend the low one down to 0 for boolean vars)
* This way, both plots span the same multiplicative range of values, and you can visualise e.g. traces of ~5 and ~500 in the same context, without one scaling the other to a line, and while preserving engineering prefixes.

(I did something similar already in the past, working with realtime data from `µ` to `k` range, so I've spent some thought on how to visualise this, although did not get around to implementing the developed approach described above)

In the OMEdit context, the "low" and "high" could actually be the left and right y-axis, and thus live in the same plot. The legends of the traces compactly track the respective engineering prefix, so no information is lost or relies on a tooltip.

**Reply by bilderbuchi** [reply]
Python prototype showing the concept (without polish!):
Successfully displays traces of nano-, milli-, one, -kilo magnitude in one plot
![image](https://github.com/OpenModelica/OpenModelica/assets/327442/1fb731d8-e099-4394-9c9d-9b8a0be0897a)

Of course there are details to deal with, e.g. plots around 30 magnitude, how to indicate "home" axis, tick alignment,...

The interested reader can surely easily come up with scenarios spanning several magnitudes from their own domain...

If one only sorts traces into "small" (-30..+30) and "large" (-1000..+1000)  buckets and leaves the axis limits auto-scaled, the results are even better, as less space is wasted in general.

**Reply by ceraolo** [reply]
Regarding your legend, in addition to @casella's comment https://github.com/OpenModelica/OpenModelica/discussions/11200#discussioncomment-7035938 that I agree with, 
 I would add that use of square brakets for units of measure (with possible presence of prefixes), although very common, is not what GCPM (General conference of weights and measures) recommends: they recommend round brackets. That's why in the present implementation OM uses round brackets.

**Reply by bilderbuchi** [reply]
Sure, as I said this was a prototype, so I did not invest time in getting details like GCPM conformance right. ;-)

re the legend -- "100 kilo" was just the name of the variable, I could have also named them A,B,C ... I'll change that.

**Reply by bilderbuchi** [reply]
I've filed #11218 with a description of this idea, so it doesn't get lost in here. Please continue there.

**Comment by s-celles** [comment]
I'm not a big fan of normalization with my students (even if I understand it makes sense and so this is probaly a feature that will have users).

Having 2 axis will solve the problem when plotting 2 variables (not more) but for such an example it will perfectly works

I also think this is a feature that will be used by some users.

On the other side I still don't understand why using scientific notation with prefix on legend won't be an acceptable solution (despide being a "poor man" solution). For most cases it's simply enough and is possibly the simplest approach to implement.

**Comment by casella** [comment, marked as answer]
> On the other side I still don't understand why using scientific notation with prefix on legend won't be an acceptable solution (despide being a "poor man" solution). For most cases it's simply enough and is possibly the simplest approach to implement.

Now I understand what you mean by that, and it makes sense. Of course the "problem" with prefixes is that their granularity is three orders of magnitude, which is too large - scientific notation is just one o.o.m.

Can you please open a ticket with your proposal (including example plots)? Then, we can discuss it with @ceraolo (which also has very strong opinions on this topic, having written a plotting tool himself), agree on a specification and then implement it.

**Reply by bilderbuchi** [reply]
> granularity is three orders of magnitude, which is too large - scientific notation is just one o.o.m.

see https://github.com/OpenModelica/OpenModelica/discussions/11200#discussioncomment-7032804 (and the comment above that) for a possible solution for this even with engineering notation/prefixes (which might be more intuitive for many people).

**Reply by casella** [reply]
I find reading it a bit awkward. The purple curve has a legend "100 kilo [kV]", does 15 mean 15 times 100 kV, i.e. 1.5 MV?

**Reply by bilderbuchi** [reply]
answered in-thread. purple is attached to the right scale (as indicated in the legend), and is about 400-600 kV

**Reply by casella** [reply]
OK, now it's much clearer. Maybe having two separate legends for the left and right scales would be even more intuitive.

**Reply by bilderbuchi** [reply]
> Maybe having two separate legends for the left and right scales would be even more intuitive.

That's an option _if_ you have the space for it. Considering OMEdits plot layout, maybe two separate rows of entries, with some indicator which belongs to what axis.
Or having solid lines for one axis, and dashed for the other. 
Or little Unicode indicator arrows in the legend entries (more compact than faking it with <-/->)
Or arrows in the plots (the way it's done in papers), but getting the placement right automatically will be tricky.
Or ...

**Reply by bilderbuchi** [reply]
I filed #11217 so that the two-axis plot thing does not get lost in here.

**Comment by s-celles** [comment]
Issue opened https://github.com/OpenModelica/OpenModelica/issues/11213

**Comment by ceraolo** [comment]
> You should notice that km/s is really not an appropriate unit here !

Maybe it is a matter of taste, but for me saying 12km/s is better than 12000 m/s or 11.2e4m/s.
 If one reads 12000 m/s or 1.2e4m/s they will mentally convert them to 12 km/s: better is to say the result directly.

Obviously, in https://github.com/OpenModelica/OpenModelica/discussions/11200#discussion-5641008 you selected a very special case in which the number is 0.003 km/s, that would have been better indicated as 3 m/s, but this is another story.  Saying 0.003 km/s is still better than 0.003 * 1e3 m/s

**Reply by casella** [reply]
The problem in that case is that the same prefix (k) was selected for both variables, which is obviously inappropriate.
